@xen-orchestra/proxy/src/app/_disposable.js

```diff
@@ -0,0 +1,36 @@
+import defer from 'promise-toolbox/defer'
+import Resource from 'promise-toolbox/_Resource'
+
+async function helper(gen, resolve, disposers) {
+  let resourceValue
+  while (true) {
+    const { value } = await gen.next(resourceValue)
+    if (value instanceof Resource) {
+      resourceValue = await value.p
+      disposers.push(value.d)
+    } else {
+      return resolve(value)
+    }
+  }
+}
+
+// inspired by https://github.com/tc39/proposal-explicit-resource-management
+export const disposable = gen => {
```

Comment:
I'm sorry about this, but I changed my mind again, I believe it should be function decorator.

I don't believe there will be cases where we want to decorate a plain generator.